butte-build/src/ir/types.rs

```diff
@@ -0,0 +1,518 @@
+            _ => false,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, TypedBuilder)]
+pub struct Enum<'a> {
+    pub ident: DottedIdent<'a>,
+    pub values: Vec<EnumVal<'a>>,
+    pub base_type: EnumBaseType,
+
+    #[builder(default)]
+    pub doc: ast::Comment<'a>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum EnumBaseType {
+    Byte,
+    UByte,
+    Short,
+    UShort,
+    Int,
+    UInt,
+    Long,
+    ULong,
+    Int8,
+    UInt8,
+    Int16,
+    UInt16,
+    Int32,
+    UInt32,
+    Int64,
+    UInt64,
+}
+
+impl<'a> std::convert::TryFrom<&ast::Type<'a>> for EnumBaseType {
+    type Error = ();
+    fn try_from(ty: &ast::Type<'a>) -> Result<Self, Self::Error> {
+        Ok(match ty {
+            ast::Type::Byte => EnumBaseType::Byte,
+            ast::Type::UByte => EnumBaseType::UByte,
+            ast::Type::Short => EnumBaseType::Short,
+            ast::Type::UShort => EnumBaseType::UShort,
+            ast::Type::Int => EnumBaseType::Int,
+            ast::Type::UInt => EnumBaseType::UInt,
+            ast::Type::Long => EnumBaseType::Long,
+            ast::Type::ULong => EnumBaseType::ULong,
+            ast::Type::Int8 => EnumBaseType::Int8,
+            ast::Type::UInt8 => EnumBaseType::UInt8,
+            ast::Type::Int16 => EnumBaseType::Int16,
+            ast::Type::UInt16 => EnumBaseType::UInt16,
+            ast::Type::Int32 => EnumBaseType::Int32,
+            ast::Type::UInt32 => EnumBaseType::UInt32,
+            ast::Type::Int64 => EnumBaseType::Int64,
+            ast::Type::UInt64 => EnumBaseType::UInt64,
+            _ => return Err(()),
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, TypedBuilder)]
+pub struct Union<'a> {
+    pub ident: DottedIdent<'a>,
+    // Ident of the enum "companion"
+    pub enum_ident: DottedIdent<'a>,
+    pub variants: Vec<UnionVariant<'a>>,
+
+    #[builder(default)]
+    pub doc: ast::Comment<'a>,
+}
+
+/// An RPC service.
+#[derive(Debug, Clone, PartialEq, TypedBuilder)]
+pub struct Rpc<'a> {
+    pub ident: DottedIdent<'a>,
+    pub methods: Vec<RpcMethod<'a>>,
+
+    #[builder(default)]
+    pub doc: ast::Comment<'a>,
+}
+
+/// A method in an RPC service.
+#[derive(Debug, Clone, PartialEq, TypedBuilder)]
+pub struct RpcMethod<'a> {
+    /// The name of the method.
+    pub ident: Ident<'a>,
+
+    /// The snake name of the method.
+    pub snake_ident: Ident<'a>,
+
+    /// The request type of the method.
+    pub request_type: DottedIdent<'a>,
+
+    /// The response type of the method.
+    pub response_type: DottedIdent<'a>,
+
+    /// Method metadata.
+    #[builder(default)]
+    pub metadata: RpcMethodMetadata,
+
+    #[builder(default)]
+    pub doc: ast::Comment<'a>,
+}
+
+#[derive(Clone, Copy, Debug, Default, PartialEq, TypedBuilder)]
+pub struct RpcMethodMetadata {
+    pub streaming: RpcStreaming,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RpcStreaming {
+    None,
+    Client,
+    Server,
+    Bidi,
```

Comment:
Is it called `Bidi` elsewhere too? If not I'd prefer just writing out `Bidirectional`.

Reply:
It's called `"bidi"` in the flatbuffers RPC spec, and I'd be happy to call it either way, but Rust usually has shorthands (e.g `Arc` or `Vec`) so I found `Bidi` a good match.

Reply:
If it’s `bidi` in the spec let’s leave `Bidi` here :)